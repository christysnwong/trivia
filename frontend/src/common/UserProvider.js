import React, {useEffect, useState} from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  stats: {
    level: 14,
    title: "Pro",
    "quizzes completed": 34,
    points: 3440,
    remainingPts: 60,
    levelPts: 300,
  },
};

const UserProvider = ({
  children,
  currUser = demoUser
}) => {
  
  const [currentUser, setCurrUser] = useState();

  return (
    // <UserContext.Provider value={{ currUser }}>
    <UserContext.Provider value={{ currUser, setCurrUser }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
