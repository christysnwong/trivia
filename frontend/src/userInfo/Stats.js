import { useState, useEffect, useContext } from "react";
import TriviaApi from "../api/api";
import UserContext from "../common/UserContext";

const Stats = () => {
  const [stats, setStats] = useState({});
  const [badges, setBadges] = useState([]);
  const { currUser } = useContext(UserContext);

  // get user's stats and badges
  useEffect(() => {
    TriviaApi.getStats(currUser.username).then((stats) => {
      setStats(stats);
    });

    TriviaApi.getBadges(currUser.username).then((badges) => {
      setBadges(badges);
    });
  }, [currUser]);


    
  return (
    <div className="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4">
      <h3 className="mb-5">
        <i className="fa-solid fa-id-badge me-2"></i>
        Statistics
      </h3>

      {Object.entries(stats)
        .filter(
          (entry) => entry[0] !== "remainingPts" && entry[0] !== "levelPts"
        )
        .map((entry, idx) => (
          <div className="row mb-3" key={idx}>
            <div className="col-4 text-capitalize fw-bold">{entry[0]}:</div>
            <div className="col-8">{entry[1]}</div>
          </div>
        ))}

      <div className="row mb-3">
        <div className="col-4">
          <b>Remaining Points to Next Level:</b>
        </div>
        <div className="col-8">{stats.remainingPts}</div>
      </div>

      <div className="row mb-5">
        <div className="col-4 fw-bold">Progress:</div>
        <div className="col-6">
          <div className="progress">
            <div
              className="progress-bar bg-warning"
              role="progressbar"
              style={{
                width:
                  ((stats.levelPts - stats.remainingPts) / stats.levelPts) *
                    100 +
                  "%",
              }}
              aria-valuenow={stats.remainingPts}
              aria-valuemin="0"
              aria-valuemax={stats.levelPts}
            ></div>
          </div>
        </div>
      </div>

      <h3 className="mb-5">
        <i className="fa-solid fa-award me-2"></i>
        Badges
      </h3>

      {!badges.length &&
        "You currently have no badges. Try to get one by doing trivia quizzes!"}

      {badges.length > 0 && (
        <div className="row mb-3">
          <div className="col-12">
            {badges.map((badge) => (
              <img
                key={badge.badgeName}
                src={badge.badgeUrl}
                alt={badge.badgeName}
                className="me-3"
              />
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default Stats;
