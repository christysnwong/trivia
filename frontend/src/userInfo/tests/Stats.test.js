import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import Stats from "../Stats";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../../common/UserProvider";

import TriviaApi from "../../api/api";

import { commonBeforeEach, commonAfterAll } from "./_testCommonUserInfo";

beforeEach(commonBeforeEach);
afterAll(commonAfterAll);

it("renders without crashing", function () {
  render(
    <UserProvider>
      <Stats />
    </UserProvider>
  );
});

it("matches snapshot", async function () {
  jest.spyOn(React, "useEffect").mockImplementation((f) => f());

  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <Stats />
      </UserProvider>
    </MemoryRouter>
  );
  

  expect(asFragment()).toMatchSnapshot();
});

it("checks if results from TriviaApi.getStats are fetched and displayed", async () => {
  jest.spyOn(React, "useEffect").mockImplementation((f) => f());
  jest.spyOn(TriviaApi, "getStats");
  render(
    <UserProvider>
      <Stats />
    </UserProvider>
  );

  
  expect(TriviaApi.getStats).toHaveBeenCalled();
  
  expect(await screen.findByText(/level/i)).toBeInTheDocument();
  expect(await screen.findByText(1)).toBeInTheDocument();
  expect(await screen.findByText(/title/i)).toBeInTheDocument();
  expect(await screen.findAllByText(/pro/i)).toHaveLength(2);
  expect(await screen.findAllByText(/points/i)).toHaveLength(2);
  expect(await screen.findByText(3440)).toBeInTheDocument();
  

});

it("checks if results from TriviaApi.getBadges are fetched and displayed", async () => {

  jest.spyOn(React, "useEffect").mockImplementation((f) => f());
  jest.spyOn(TriviaApi, "getBadges");
  render(
    <UserProvider>
      <Stats />
    </UserProvider>
  );

  expect(TriviaApi.getBadges).toHaveBeenCalled();

  const badgePro = await screen.findByAltText(/pro/i);
  const badgeApp = await screen.findByAltText(/apprentice/i);
  const badgeNewbie = await screen.findByAltText(/newbie/i);
  
  expect(badgePro).toBeInTheDocument();
  expect(badgeApp).toBeInTheDocument();
  expect(badgeNewbie).toBeInTheDocument();

  

});