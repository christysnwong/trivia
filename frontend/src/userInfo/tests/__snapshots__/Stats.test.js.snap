// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4"
  >
    <h3
      class="mb-5"
    >
      <i
        class="fa-solid fa-id-badge me-2"
      />
      Statistics
    </h3>
    <div
      class="row mb-3"
    >
      <div
        class="col-4"
      >
        <b>
          Remaining Points to Next Level:
        </b>
      </div>
      <div
        class="col-8"
      />
    </div>
    <div
      class="row mb-5"
    >
      <div
        class="col-4 fw-bold"
      >
        Progress:
      </div>
      <div
        class="col-6"
      >
        <div
          class="progress"
        >
          <div
            aria-valuemin="0"
            class="progress-bar bg-warning"
            role="progressbar"
          />
        </div>
      </div>
    </div>
    <h3
      class="mb-5"
    >
      <i
        class="fa-solid fa-award me-2"
      />
      Badges
    </h3>
    You currently have no badges. Try to get one by doing trivia quizzes!
  </div>
</DocumentFragment>
`;
